Framework/Algorithms/src/CalcCountRate.cpp

```diff
@@ -0,0 +1,653 @@
+    return;
+  }
+
+  if (m_XRangeMin == EMPTY_DBL()) {
+    m_XRangeMin = realMin;
+  }
+  if (m_XRangeMax == EMPTY_DBL()) {
+    // include rightmost limit into the histogramming
+    m_XRangeMax = realMax * (1. + std::numeric_limits<double>::epsilon());
+  }
+  if (m_XRangeMin < realMin) {
+    g_log.debug() << "Workspace constrain min range changed from: "
+                  << m_XRangeMin << " To: " << realMin << std::endl;
+    m_XRangeMin = realMin;
+  }
+  if (m_XRangeMax > realMax) {
+    g_log.debug() << "Workspace constrain max range changed from: "
+                  << m_XRangeMax << " To: " << realMax << std::endl;
+    m_XRangeMax = realMax * (1. + std::numeric_limits<double>::epsilon());
+  }
+}
+
+/**Check if visualization workspace is necessary and initiate it if requested.
+* Sets or clears up internal m_visWS pointer and "do-visualization workspace"
+* option.
+*/
+void CalcCountRate::checkAndInitVisWorkspace() {
+  std::string visWSName = getProperty("VisualizationWs");
+  if (visWSName.empty()) {
+    m_visWs.reset();
+    m_doVis = false;
+    return;
+  }
+  m_doVis = true;
+
+  int numTBins = getProperty("NumTimeSteps");
+  if (this->notmalizeCountRate()) {
+    if (numTBins > m_numLogSteps) {
+      g_log.information() << "Number of time step in normalized visualization "
+                             "workspace exceeds the number of points in the "
+                             "normalization log. This mode is not supported so "
+                             "number of time steps decreased to be equal to "
+                             "the number of normalization log points\n";
+      numTBins = m_numLogSteps;
+    }
+  }
+  int numXBins = getProperty("XResolution");
+  std::string RangeUnits = getProperty("RangeUnits");
+
+  m_visWs = boost::dynamic_pointer_cast<DataObjects::Workspace2D>(
+      API::WorkspaceFactory::Instance().create("Workspace2D", numTBins,
+                                               numXBins + 1, numXBins));
+  m_visWs->setTitle(visWSName);
+
+  double Xmax = m_XRangeMax;
+  // a bit dodgy code. It can be generalized
+  if (std::isinf(Xmax)) {
+    auto Xbin = m_workingWS->readX(0);
+    for (int64_t i = Xbin.size() - 1; i >= 0; --i) {
+      if (!std::isinf(Xbin[i])) {
+        Xmax = Xbin[i];
+        break;
+      }
+    }
+    if (std::isinf(Xmax)) {
+      g_log.warning() << "All X-range for visualization workspace is infinity. "
+                         "Can not build visualization workspace in the units "
+                         "requested\n";
+      m_visWs.reset();
+      m_doVis = false;
+      return;
+    }
+  }
+
+  // define X-axis in target units
+  double dX = (Xmax - m_XRangeMin) / numXBins;
+  std::vector<double> xx(numXBins);
+  for (int i = 0; i < numXBins; ++i) {
+    xx[i] = m_XRangeMin + (0.5 + static_cast<double>(i)) * dX;
+  }
+  auto ax0 = new API::NumericAxis(xx);
+  ax0->setUnit(RangeUnits);
+  m_visWs->replaceAxis(0, ax0);
```

Comment:
To make the ownership change crystal clear, you could use
```c++
auto ax0 = make_unique<API::NumericAxis>(xx);
ax0->setUnit(RangeUnits);
m_visWs->replaceAxis(0, ax0.release());
```

You can also reuse `ax0` below instead of introducing `ax1`.